File: software/fizz-buzz/fb.ts
function assert(condition: boolean, msg?: string): asserts condition {
	if (!condition)
		throw new Error(msg)
}

function is_multiple(x: number, m: number): boolean
function is_multiple(x: bigint, m: bigint): boolean
function is_multiple(x: number | bigint, m: number | bigint) {
	return m == 0 ?
		x == 0 :
		(x as any) % (m as any) == 0
}

// tests
assert(is_multiple(0, 0))
assert(is_multiple(0, 1) && !is_multiple(1, 0))
assert(is_multiple(2, 1) && !is_multiple(1, 2))
assert(is_multiple(4, 2) && !is_multiple(2, 4))
assert(is_multiple(9, 3) && !is_multiple(3, 9))

function* range(n: bigint, end?: bigint) {
	for (; end === undefined ? Infinity : n < end; n++)
		yield n
}

const generic_fizzbuzz = (cases: Map<bigint, string>) =>
	range(0n).map(i => [...cases.entries()
		.filter(([k]) => is_multiple(i, k))
		.map(([, v]) => v)
	].join('') || i)

console.log([...generic_fizzbuzz(
	new Map([[3n, "Fizz"], [5n, "Buzz"], [7n, "Bazz"]] as const)
).take(3 * 5 * 7 + 1)])

File: software/fizz-buzz/fb.py
#!/usr/bin/env python3
from typing import Final, List, Dict
from sys import argv

def is_multiple(x: float | int, m: float | int, /):
	return x == 0 if m == 0 else x % m == 0

# tests
assert is_multiple(0,0)
assert is_multiple(0,1) and not is_multiple(1,0)
assert is_multiple(2,1) and not is_multiple(1,2)
assert is_multiple(4,2) and not is_multiple(2,4)
assert is_multiple(9,3) and not is_multiple(3,9)

def generic_fizzbuzz(n: int, cases: Dict[int, str]):
	a: List[int | str] = []
	for i in range(0, n):
		a.append(''.join(v for k, v in cases.items() if is_multiple(i, k)) or i)
	return a

def main(*args: str):
	if len(args) > 0:
		n = int(args[0])
	else:
		n = 3*5*7 + 1
	print(generic_fizzbuzz(n, {3: "Fizz", 5: "Buzz", 7: "Bazz"}))

if __name__ == '__main__':
	main(*argv[1:])
